Priscila Cunha Vazquez RA: 11.121.322-9

O objetivo do código é ter um classificador para identificar, de acordo com as características de entrada, se há ou não câncer de mama.
Como os dados de entrada são contínuos é necessário utilizar o método GaussianNB do Naive Bayes, visto que a probabilidade condicional será obtida através de uma gaussiana.

Para isso, foi utilizado o banco de dados de câncer de mama do sklearn.
Então, primeiramente foi importado o método e o banco de dados, separando as entradas por característica.

In [65]:
# Importa do sklearn naive bayes o método GaussianNB, já que os dados de entrada são contínuos
from sklearn.naive_bayes import GaussianNB

#dados utilizados do banco de dados
from sklearn.datasets import load_breast_cancer

X, y = load_breast_cancer(return_X_y = True)

#entrada
raio = X[:, 1]
textura = X[:, 2]
perimetro = X[:, 3]
area = X[:, 4]
suavidade = X[:, 5]
compactacao = X[:, 6]
concavidade = X[:, 7]
ptos_concavos = X[:, 8]
simetria = X[:, 9]
dimensao_fractal = X[:, 10]

#saida
target = y[:]

Em seguida, foi utilizada a função GaussianNB.fit para aplicar o Naive Bayes à todos os dados de entrada e saída.

In [66]:
nb = GaussianNB()
nb.fit(X, target)

GaussianNB()

Depois foi utilizado o score, que retorna a precisão média dos dados e rótulos fornecidos. Ela é utilizada para ver a acurácia - quanto mais perto de 1, melhor a acurácia do modelo.

In [67]:
nb.score(X, target)

0.9420035149384886

Para testar o modelo, foi feita uma separação onde uma parte do modelo será utilizada para treinamento e a outra parte para testes. 

In [68]:
from sklearn.model_selection import train_test_split

# separa os dados em treino e teste
entrada_treino, entrada_teste, saida_treino, saida_teste = train_test_split(X, target, test_size=0.3)

# cria o modelo de classificação e treina
nb_model = GaussianNB()
nb_model.fit(entrada_treino, saida_treino)

GaussianNB()

Com as entradas da parte de teste do modelo, foi utilizado o predict() do modelo Gaussiano, que serve para prever qual seria a saída de acordo com as entradas fornecidas. Assim, nesse caso ele esta prevendo se a pessoa tem ou não cancer de mama de acordo com as caracteristicas.

In [74]:
predictions = [nb.predict([entry]) for entry in entrada_teste]

print(predictions)

[array([1]), array([0]), array([1]), array([1]), array([0]), array([1]), array([1]), array([1]), array([1]), array([1]), array([1]), array([1]), array([1]), array([1]), array([1]), array([0]), array([1]), array([1]), array([0]), array([0]), array([1]), array([1]), array([1]), array([1]), array([1]), array([0]), array([1]), array([1]), array([0]), array([1]), array([1]), array([1]), array([0]), array([1]), array([1]), array([1]), array([1]), array([0]), array([1]), array([0]), array([1]), array([1]), array([0]), array([0]), array([1]), array([0]), array([0]), array([1]), array([1]), array([0]), array([1]), array([1]), array([0]), array([1]), array([0]), array([0]), array([1]), array([1]), array([0]), array([1]), array([0]), array([1]), array([1]), array([0]), array([1]), array([1]), array([1]), array([1]), array([1]), array([0]), array([1]), array([0]), array([1]), array([1]), array([0]), array([0]), array([1]), array([1]), array([1]), array([1]), array([0]), array([1]), array([0]), arr

Já para o predict_proba ele fornece as probabiliades de ter dado positivo ou negativo para o cancer de mama para cada caso.

In [75]:
predictions = [nb.predict_proba([entry]) for entry in entrada_teste]

print(predictions)

[array([[8.12064124e-16, 1.00000000e+00]]), array([[1.00000000e+00, 9.11892508e-24]]), array([[0.01621676, 0.98378324]]), array([[1.25607339e-09, 9.99999999e-01]]), array([[0.86552706, 0.13447294]]), array([[4.53852431e-12, 1.00000000e+00]]), array([[3.40956422e-19, 1.00000000e+00]]), array([[5.3897919e-15, 1.0000000e+00]]), array([[4.20239878e-14, 1.00000000e+00]]), array([[3.29537002e-08, 9.99999967e-01]]), array([[9.9423948e-15, 1.0000000e+00]]), array([[0.00940411, 0.99059589]]), array([[6.77175147e-11, 1.00000000e+00]]), array([[5.50915156e-14, 1.00000000e+00]]), array([[5.04387661e-16, 1.00000000e+00]]), array([[1.00000000e+00, 6.13616025e-33]]), array([[5.95654255e-14, 1.00000000e+00]]), array([[4.60338559e-16, 1.00000000e+00]]), array([[1.00000000e+00, 1.46465341e-51]]), array([[1.00000000e+00, 4.87429722e-86]]), array([[9.87409319e-14, 1.00000000e+00]]), array([[3.71186925e-07, 9.99999629e-01]]), array([[1.55661861e-07, 9.99999844e-01]]), array([[9.10331815e-15, 1.00000000e+00

Utilizando o classification report, é possível analisar a eficácia do treinamento - já que uma parte do modelo foi usada para treinar, e a outra parte foi feita de teste. Essa parte de teste tem sua análise completa nesse feedback.

In [69]:
from sklearn.metrics import classification_report

# realiza a predição e imprime o relatório de classificação
print(classification_report(saida_teste, nb_model.predict(entrada_teste)))

              precision    recall  f1-score   support

           0       0.93      0.85      0.89        66
           1       0.91      0.96      0.94       105

    accuracy                           0.92       171
   macro avg       0.92      0.91      0.91       171
weighted avg       0.92      0.92      0.92       171



A partir da análise dos resultados dos testes é possível concluir que o modelo teve um bom treinamento, com uma precisão de 93% de acerto para resultados de câncer negativos, e 91% de precisão de acerto para câncer de mama positivo. A acurácia do modelo era de 0.94, e com o treinamento a acurácia foi de 0.92 - assim, ela foi reduzida, mas o modelo ainda funciona na maior parte dos teste.